### PR TITLE
Implemented TypeSerializer for the argument/configuration system.

### PR DESCRIPTION
**Pull request by orbyfied**

### CODE REVIEW

File: src/main/java/com/domain/redstonetools/features/AbstractFeature.java [hunk 1]
```diff
@@ -3,11 +3,41 @@
...
+    final Feature featureDesc;
+
+    {
+        featureDesc = getClass().getAnnotation(Feature.class);
+        if (featureDesc == null)
+            throw new IllegalStateException("Supposed feature " + getClass() + " does not" +
+                    " have an @Feature descriptor annotation");
+    }
+
+    public Feature getFeatureDescriptor() {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/AbstractFeature.java:22]: I think we might as well get rid of this and add the other fields as getters

File: src/main/java/com/domain/redstonetools/features/arguments/Argument.java [hunk 2]
```diff
@@ -39,18 +40,19 @@ public Argument<T> ensureNamed(String fieldName) {
         return this;
     }
 
-    public String getName() {
+    public String name() {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/Argument.java:43]: No need to rename this, getName follows conventions and is clearer

File: src/main/java/com/domain/redstonetools/features/arguments/Argument.java [hunk 3]
```diff
@@ -39,18 +40,19 @@ public Argument<T> ensureNamed(String fieldName) {
...
         return name;
     }
 
-    public ArgumentType<T> getType() {
+    public TypeSerializer<T> type() {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/Argument.java:47]: same goes for this

File: src/main/java/com/domain/redstonetools/features/arguments/BlockColorSerializer.java [hunk 4]
```diff
@@ -6,12 +6,16 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
-public class BlockColorArgumentType extends SetArgumentType<BlockColor> {
+public class BlockColorSerializer extends OptionSetSerializer<BlockColor> {
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/arguments/BlockColorSerializer.java:9]: Could you remove OptionSetSerializer and create a generic EnumSerializer instead?

File: src/main/java/com/domain/redstonetools/features/Feature.java [hunk 5]
```diff
@@ -8,6 +8,7 @@
 @Target(ElementType.TYPE)
 @Retention(RetentionPolicy.RUNTIME)
 public @interface Feature {
+    String id(); // internal name
```
Comment (Matthias1590) [inline src/main/java/com/domain/redstonetools/features/Feature.java:11]: Please create a separate branch and PR for this, it's not related to argument types

**Review by Matthias1590 — CHANGES_REQUESTED** [review]

File: src/main/java/com/domain/redstonetools/features/Feature.java (hunk 5 above)
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/features/Feature.java:11]: Yeah idk why it was brought over from another branch but removing it is a pain.

**Comment by Matthias1590** [comment]
I'll create an issue for the remaining 2 comments and merge for now

**Pull request merged**